h2/src/main/org/h2/expression/Comparison.java

```diff
@@ -272,7 +272,7 @@ public Value getValue(Session session) {
         }
         int dataType = Value.getHigherOrder(left.getType(), right.getType());
         if (dataType == Value.ENUM) {
-            String[] enumerators = getEnumerators(l, r);
+            String[] enumerators = ValueEnum.getEnumeratorsForBinaryOperation(l, r);
             l = l.convertToEnum(enumerators);
```

Comment:
hmmm, looking at the code here I wonder what the expected behaviour is when the enums are different, but the enumerators are the same

Reply:
You mean that someone could compare values from the different columns with different data types? We may have a problem here.

Reply:
Yup, that's what I'm thinking. We'll probably find that MySQL allows it and just does a string comparison of the enum names

Reply:
Our storage format for ENUM values is probably defective by design. When an `ENUM` value is read from the PageStore or MVStore tables and indexes can't even check that this is enum for the first 16 constants. This is one of the reasons of some existing hacks. Also H2 doesn't have a centalized place for proper recreation of the stored values for values that don't have an exact data type in the storage. Indexes, for example, use this method too in weird places.

I'm not sure that we can resolve all the issues with `ENUM` data type without breaking of compatibility in storage formats.

Reply:
Our reuslt is weird, an empty result set is expected here. But this is not changed in this pull request.
```SQL
CREATE TABLE TEST(ID INT, E1 ENUM('A', 'B', 'C'), E2 ENUM('C', 'D', 'E'));
INSERT INTO TEST VALUES (1, 'A', 'C'), (2, 'C', 'C');

SELECT * FROM TEST WHERE E1 = E2;
> ID E1 E2
> -- -- --
> 1  A  C
> rows: 1
```

Reply:
No, H2 is not compatible with MySQL. And is not compatible with PostgreSQL too.

PostgreSQL forbids such comparison between different ENUM types.
MySQL allows it and compares its string representations.
H2 allows it and compares its ordinal values.

Reply:
So it's not easy to say what we should expect here.

Reply:
Build failures are probably not related.

Reply:
Probably we can't change comparison operation to be MySQL-compatible without serious changes in different places (may be even in storage format). Comparison is used even when exact type is not really known, during scanning of indexes, for example. Storage format is `INT`-based, so we need comparisons by ordinal values.

However, we can block comparison between `ValueEnum` instanced wih different enumeratiors (like PostgreSQL). But I'm not sure that such restriction will be reliable. This method can be invoked sometimes with `ValueString` or `ValueInt` as one of its arguments if value was read directly from the storage. This trick is here due to this problem. In this case we can't detect types of both values properly.

Reply:
Yeah, when I implemented Enum I didn't think all these cases through. I should really have given it another identifier in the storage format.
Possibly we have to do the painful thing and have another flag for storing it properly and require people to do a dump and restore when switching to using the flag.

Given that there is no widespread conformity in other projects, I'm not terribly worried about being compatible.

I'm OK with blocking silliness only in places where it is easy for us.

Reply:
OK, I restarted the build (due to PageStore testing issue).

This pull request is not changing anything with comparison, so from my point of view it can be merged after testing. At least we will have one more test for a specific operation (I cannot reproduce this problem from a pure SQL script with the same commands for some unknown reason.)

And I agree that we need a better generic solution for a new databases with sane logic not scattered around different places.